plugins/com.google.cloud.tools.eclipse.projectselector.test/src/com/google/cloud/tools/eclipse/projectselector/MiniSelectorTest.java

```diff
@@ -89,31 +101,45 @@ public void testListenerFired() {
 
     assertNotNull(selector.getSelection());
     assertTrue(selector.getSelection().isEmpty());
-    final boolean[] calledAndCorrect = new boolean[] {false};
+    final AtomicReference<ISelection> lastSelection = new AtomicReference<>();
     selector.addSelectionChangedListener(new ISelectionChangedListener() {
       @Override
       public void selectionChanged(SelectionChangedEvent event) {
-        assertThat(event.getSelection(), instanceOf(IStructuredSelection.class));
-        IStructuredSelection selection = (IStructuredSelection) event.getSelection();
-        assertEquals(1, selection.size());
-        assertThat(selection.getFirstElement(), instanceOf(GcpProject.class));
-        GcpProject gcpProject = (GcpProject) selection.getFirstElement();
-        calledAndCorrect[0] =
-            "foo".equals(gcpProject.getName()) && "foo.id".equals(gcpProject.getId());
+        lastSelection.set(event.getSelection());
       }
     });
-    assertFalse(calledAndCorrect[0]);
 
     selector.setProject("foo.id");
-    spinEvents(); // setProject() waits for the project list to be returned
+    waitUntilResolvedProject(selector); // waits for the project list to be returned
 
-    assertTrue(calledAndCorrect[0]);
+    assertThat(lastSelection.get(), instanceOf(IStructuredSelection.class));
+    IStructuredSelection selection = (IStructuredSelection) lastSelection.get();
+    assertEquals(1, selection.size());
+    assertThat(selection.getFirstElement(), instanceOf(GcpProject.class));
+    GcpProject gcpProject = (GcpProject) selection.getFirstElement();
+    assertEquals("foo", gcpProject.getName());
+    assertEquals("foo.id", gcpProject.getId());
   }
 
-  private void spinEvents() {
-    while (Display.getCurrent().readAndDispatch());
+  private void waitUntilResolvedProject(final MiniSelector selector) {
+    bot.waitUntil(new DefaultCondition() {
+      @Override
+      public boolean test() throws Exception {
+        if (Display.getCurrent() != null) {
```

Comment:
This seems weird. We are in the UI thread, we already created `MiniSelector` and went through a lot of SWT methods at this point. This can never be null. I think we can remove it.

Reply:
The nice thing with using SWTBot is that it doesn't matter whether the test was run in the UI thread or a different thread (tycho-surefire `<useUIThread>`).